### ISSUE TITLE
Pod with XCFrameworks embedded libraries not compiling for "Generic iOS Device"

### ISSUE THREAD

**Issue by remizorrr** [CocoaPods/CocoaPods#9794]
Setup is:
Xcode 11.4.1,
Cocapods top of tree master.

Local pod, with embedded (vendored_frameworks) xcframeworks.
the project, referencing local pod.

compiling and running on the iPhone - everything works.
compiling and running on the Simulator - everything works.
Compiling for "Generic iOS Device" - Shows "No such module 'FrameworkName'"
for the embedded XCFramework.

A project with same xcframework, embedded using Xcode compiles fine.

Example projects attached:
[Project-Not-Working.zip](https://github.com/CocoaPods/CocoaPods/files/4640013/Project-Not-Working.zip)
[Project-Working.zip](https://github.com/CocoaPods/CocoaPods/files/4640014/Project-Working.zip)

related to the previously closed issue: https://github.com/CocoaPods/CocoaPods/issues/9785

**Comment by dnkoutso** [comment]
Did you try cocoapods master?

**Comment by remizorrr** [comment]
Yes. the issue is on master.

**Comment by remizorrr** [comment]
@dnkoutso any updates?

**Comment by dnkoutso** [comment]
No updates here, I am sorry I do not currently have the time to look into this. Best to investigate further yourself and perhaps make a PR so we can fix it for everyone if its an issue.

**Comment by remizorrr** [comment]
@dnkoutso I had a peak in the code. It will take a lot of time for me to find a cause of the issue. Is there anyone that could get it done? Looks like I am no the only one having issues with xcframeworks.

**Comment by amorde** [comment]
We have a lot of changes to xcframeworks that will be released in the next version, but since you tried master would need to investigate further.

Personally have been taking a bit of a break from CocoaPods and have started to fall behind on other things, so I'm not sure when I will have time to look in to this. The example projects are super helpful, thank you for including them 🙏

**Comment by remizorrr** [comment]
@amorde Thank you a lot. I will greatly appreciate if you could look into that. It is currently blocking our library consumers from deploying their apps to the app store. Let me know if you need any help. I would love to give you a hand here.

**Comment by amorde** [comment]
Off the top of my head I would check the architectures and whether it was built with the BUILD_LIBRARY_FOR_DISTRIBUTION.

What is the difference between the working and not working examples you uploaded?

**Comment by seanreinhardtapps** [comment]
The issue is that the included XCFramework has a variant slice of `arm64`; and building for Generic iOS Device builds for valid device architectures `armv7 arm64`.  Since the XCFramework doesn't contain all of the expected architectures, the `copy_xcframework_script` fails with error:
> [CP] Unable to find matching .xcframework slice in 'Project-Not-Working/Project/Pods/../../Library1/Frameworks/CTFoundation.xcframework CTFoundation framework ios-arm64/CTFoundation.framework ios-x86_64-simulator/CTFoundation.framework' for the current build architectures (armv7 arm64).

The script looks to be working as designed, since the XCF only has arm64 and the App is building for arm64 and armv7.  If we can identify any improvements though, I'm willing to take a shot at working on it.

**Comment by amorde** [comment]
Thank you for digging into this @seanreinhardtapps - yes that sounds like it's working as intended. 

If the .xcframework is lacking the architecture you are building for, you cannot build the app with that framework. You'll need to either refine `VALID_ARCHITECTURES` or add support for this architecture to your xcframework. I don't think there's anything CocoaPods can do here

**Comment by amorde** [comment]
Though I'm curious how/why Xcode builds this when using native integration. Worth looking in to

**Comment by seanreinhardtapps** [comment]
More or less because Xcode will still link the ios-arm architecture slice to the app.  The build error is ultimately observed is essentially a linking error.  The binary could still link successfully with a missing architecture, it's at runtime that you'd have a problem.  
From what I understand, Xcode decides which slice to link based solely on the entries in the XCFrameworks info.plist, and its the consumers issue if there's an architecture missing.

**Comment by amorde** [comment]
Oh that's right - does the "Project Working" project actually run on a device or does it crash with a `dyld` error? Apologies for asking these questions without having tested this myself

**Comment by seanreinhardtapps** [comment]
It launches on the arm64 device I installed it on.  It would almost certainly get a dyld error if installed on an armv7 device.

One other thing I noticed, for a Pod to link an XCFramework with only an arm64 slice, the VALID_ARCHITECTURES of the Pod target would have to be configured accordingly (the valid archs of the App do not factor into the logic of copy_xcframeworks_script).  To achieve this configuration, I think the podspec containing an XCFramework like this would need:
`s.pod_target_xcconfig = { 'VALID_ARCHS' => 'arm64' }`.

**Comment by remizorrr** [comment]
Hi @seanreinhardtapps. Thank you for looking into this issue. Basically the app, and frameworks are ios 11+ So I can't compile them under armv7 architecture, as it is deprecated.
setting `s.pod_target_xcconfig = { 'VALID_ARCHS' => 'arm64' }.` on a "umbrella" podspec didn't help so far.

**Comment by remizorrr** [comment]
@seanreinhardtapps @amorde 
So basically in case I am trying to build an xcframework for armv7, or armv7s, I get a failure.
```
warning: The armv7 architecture is deprecated for your deployment target (iOS 11.0). You should update your ARCHS build setting to remove the armv7 architecture. (in target 'CTFoundation' from project 'Pods')
** ARCHIVE FAILED **
```

**Comment by remizorrr** [comment]
Btw, after updating to latest cocoapods (top of the tree master), cleaning up all cached workspaces etc, and running pod install with `s.pod_target_xcconfig = { 'VALID_ARCHS' => 'arm64' }.` the compiler seemd to be working with no issues.

**Comment by remizorrr** [comment]
Hm. After I have removed `{ 'VALID_ARCHS' => 'arm64' }.` everything still works now on top of the tree master.
on 1.9.3 simulator build doesn't work, but "genertic" and "device" builds work fine.
Looks like the valid_archs list was modified in "1.9.3". so now it doesn't have `armv7` and `armv7s`.

**Comment by jakub-vallo** [comment]
Hello,

I have the same problem as described in the issue title.

My setup:
Tested both: Xcode 11.5 and Xcode 11.3.1
cocoapods version 1.9.3

I am getting this warning, which is causing the build to fail.

```warning: [CP] Unable to find matching .xcframework slice in ' true ios-armv7_armv7s_arm64/XXX.framework ios-i386_x86_64-simulator/XXX.framework' for the current build architectures (armv7 arm64).```

I think the issue is in ```Pods-XXX-artifacts.sh install_xcframework()```.

I have modified ```Pods-XXX-artifacts.sh install_xcframework()``` in a following way:

```
install_xcframework() {
  local basepath="$1"
  local dsym_folder="$2"
  local embed="$3"
  shift
  local paths=("$@")

  # Locate the correct slice of the .xcframework for the current architectures
  local target_path=""
  local target_arch="$ARCHS"

  # Replace spaces in compound architectures with _ to match slice format
  target_arch=${target_arch// /_}

  local target_variant=""
  if [[ "$PLATFORM_NAME" == *"simulator" ]]; then
    target_variant="simulator"
  fi
  if [[ ! -z ${EFFECTIVE_PLATFORM_NAME+x} && "$EFFECTIVE_PLATFORM_NAME" == *"maccatalyst" ]]; then
    target_variant="maccatalyst"
  fi
  for i in ${!paths[@]}; do
    echo "install_xcframework - paths: ${paths[$i]}"
    echo "install_xcframework - target_arch: $target_arch"
    if [[ "${paths[$i]}" == *"$target_arch"* ]] && [[ "${paths[$i]}" == *"$target_variant"* ]]; then
      # Found a matching slice
      echo "Selected xcframework slice ${paths[$i]}"
      target_path=${paths[$i]}
      break;
    fi
  done

  if [[ -z "$target_path" ]]; then
    echo "warning: [CP] Unable to find matching .xcframework slice in '${paths[@]}' for the current build architectures ($ARCHS)."
    return
  fi

  install_framework "$basepath/$target_path" "$embed"

  if [[ -z "$dsym_folder" || ! -d "$dsym_folder" ]]; then
    return
  fi

  dsyms=($(ls "$dsym_folder"))

  local target_dsym=""
  for i in ${!dsyms[@]}; do
    install_artifact "$dsym_folder/${dsyms[$i]}" "$CONFIGURATION_BUILD_DIR" "true"
  done
}
```

I get this in logs:

```
install_xcframework - paths: 
install_xcframework - target_arch: armv7_arm64
install_xcframework - paths: true
install_xcframework - target_arch: armv7_arm64
install_xcframework - paths: ios-armv7_armv7s_arm64/XXX.framework
install_xcframework - target_arch: armv7_arm64
install_xcframework - paths: ios-i386_x86_64-simulator/XXX.framework
install_xcframework - target_arch: armv7_arm64
warning: [CP] Unable to find matching .xcframework slice in ' true ios-armv7_armv7s_arm64/XXX.framework ios-i386_x86_64-simulator/XXX.framework' for the current build architectures (armv7 arm64).
```

I think the problem is ```armv7_arm64``` will not get matched to ```ios-armv7_armv7s_arm64``` here: ```if [[ "${paths[$i]}" == *"$target_arch"* ]] && [[ "${paths[$i]}" == *"$target_variant"* ]];```

Hope it will be helpful.

**Comment by jakub-vallo** [comment]
Not sure if you already solved this issue, but I have opened your `Project-Not-Working.zip` and the reason, why it does not build is that Library1 target in Pods project has min iOS version set to 10.0. Because of that Xcode will include armv7 in current build architectures when building for "Generic iOS Device". To remove armv7 you need to set min iOS version of Library1 target to iOS 11.0+.

The reason why it works for your real device is that your real device has probably iOS version 11.0+ and therefore Xcode will not include armv7 in current build architectures.

**Comment by dnkoutso** [comment]
Is this still an issue in CocoaPods then?

**Comment by jakub-vallo** [comment]
I believe the original issue was not related to CocoaPods, but to the wrong project setup.
The issue I have described in my comment is in fact different issue and it is fixed by #9790.
I think you can close this.

**Comment by dnkoutso** [comment]
great thanks!
